Userland/Libraries/LibSQL/SQLResult.h

```diff
@@ -0,0 +1,140 @@
+    S(SchemaExists, "Schema {} already exist")                  \
+    S(TableDoesNotExist, "Table {} does not exist")             \
+    S(TableExists, "Table {} already exist")                    \
+    S(InvalidType, "Invalid type {}")
+
+enum class SQLErrorCode {
+#undef __ENUMERATE_SQL_ERROR
+#define __ENUMERATE_SQL_ERROR(error, description) error,
+    ENUMERATE_SQL_ERRORS(__ENUMERATE_SQL_ERROR)
+#undef __ENUMERATE_SQL_ERROR
+};
+
+struct SQLError {
+    SQLErrorCode code { SQLErrorCode::NoError };
+    String error_argument { "" };
+
+    String to_string() const
+    {
+        String code_string;
+        String message;
+        switch (code) {
+#undef __ENUMERATE_SQL_ERROR
+#define __ENUMERATE_SQL_ERROR(error, description) \
+    case SQLErrorCode::error:                     \
+        code_string = #error;                     \
+        message = description;                    \
+        break;
+            ENUMERATE_SQL_ERRORS(__ENUMERATE_SQL_ERROR)
+#undef __ENUMERATE_SQL_ERROR
+        default:
+            VERIFY_NOT_REACHED();
+        }
+        if (!error_argument.is_null() && error_argument == "") {
+            if (message.find("{}").has_value()) {
+                message = String::formatted(message, error_argument);
+            } else {
+                message = String::formatted("{}: {}", message, error_argument);
+            }
+        }
+        if (message.is_null() || (message == "")) {
+            return code_string;
+        } else {
+            return String::formatted("{}: {}", code_string, message);
+        }
+    }
+};
+
+class SQLResult : public Core::Object {
+    C_OBJECT(SQLResult)
+
+public:
+    void append(Tuple const& tuple)
+    {
+        if (!m_result_set)
+            m_result_set = make<Vector<Tuple>>();
+        m_result_set->append(tuple);
+    }
+
+    SQLCommand command() const { return m_command; }
+    int updated() const { return m_update_count; }
+    int inserted() const { return m_insert_count; }
+    int deleted() const { return m_delete_count; }
+    SQLError const& error() const { return m_error; }
+    bool has_results() const { return m_result_set.ptr() != nullptr; }
+    Vector<Tuple> const* results() const { return m_result_set; }
+
+private:
+    SQLResult() = default;
+
+    explicit SQLResult(SQLCommand command, int update_count = 0, int insert_count = 0, int delete_count = 0)
+        : m_command(command)
+        , m_update_count(update_count)
+        , m_insert_count(insert_count)
+        , m_delete_count(delete_count)
+    {
+    }
+
+    SQLResult(SQLCommand command, SQLErrorCode error_code, String error_argument)
+        : m_command(command)
+        , m_error({ error_code, move(error_argument) })
+    {
+    }
+
+    SQLCommand m_command { SQLCommand::Select };
+    SQLError m_error { SQLErrorCode::NoError, "" };
+    int m_update_count { 0 };
+    int m_insert_count { 0 };
+    int m_delete_count { 0 };
+    OwnPtr<Vector<Tuple>> m_result_set { nullptr };
```

Comment:
No, because there will not always be a result set. Only SELECT statements and INSERT/UPDATE ... RETURNING do.

Reply:
Sorry, read too quickly. I am not yet familiar enough with the xxxxPtrVector family of templates. Will investigate

Reply:
Yeah `NonnullOwnPtrVector` is allowed to be empty, which I think would allow for the behavior you want. It's basically a vector of `OwnPtr` with the restriction that each `OwnPtr` entry cannot be null.

Reply:
`NonnullOwnPtrVector` == `Vector<NonnullOwnPtr>`. What you have here is a `OwnPtr` to a `Vector`, not a `Vector` of `OwnPtr`s.